pennylane/queuing.py

```diff
@@ -11,8 +11,106 @@
+
+When an object is queued, it sends itself to the :class:`~.QueuingManager`. The :class:`~.QueuingManager`
+is a global singleton class that facilitates placing objects in the queue. All of :class:`~.QueuingManager`'s methods
+and properties are class methods and properties, so all instances will access the same information.
+
+The :meth:`~.QueuingManager.active_context` is the queue where any new objects are placed.
+The :class:`~.QueuingManager` is said to be *recording* if an active context exists.
+
+Active contexts are :class:`~.AnnotatedQueue` instances. They are *context managers* where recording occurs
+within a ``with`` block.
+
+Let's take a look at an example. If we query the :class:`~.QueuingManager` outside of an
+:class:`~.AnnotatedQueue`'s context, we can see that nothing is recording and no active context exists.
+
+>>> print("Are we recording? ", qml.QueuingManager.recording())
+Are we recording?  False
+>>> print("What's the active context? ", qml.QueuingManager.active_context())
+What's the active context?  None
+
+Inside of a context, we can see the active recording context:
+
+>>> with qml.queuing.AnnotatedQueue() as q:
+...     print("Are we recording? ", qml.QueuingManager.recording())
+...     print("Is q the active queue? ", q is qml.QueuingManager.active_context())
+Are we recording?  True
+Is q the active queue?  True
+
+If we construct an operator inside the recording context, we can see it is added to the queue:
+
+>>> with qml.queuing.AnnotatedQueue() as q:
+...     op = qml.PauliX(0)
+>>> q.queue
+[PauliX(wires=[0])]
+
+If an operator is constructed outside of the context, we can manually add it to the queue by
+calling the :meth:`~.Operator.queue` method:
+
+>>> op = qml.PauliX(0)
+>>> with qml.queuing.AnnotatedQueue() as q:
+...     op.queue()
+>>> q.queue
+[PauliX(wires=[0])]
+
+An object can only exist up to *once* in the queue, so calling queue multiple times will
+not do anything.
+
+>>> op = qml.PauliX(0)
+>>> with qml.queuing.AnnotatedQueue() as q:
+...     op.queue()
+...     op.queue()
+>>> q.queue
+[PauliX(wires=[0])]
+
+The :func:`~.apply` method allows a single object to be queued multiple times in a circuit.
+The function essentially queues a copy of the original object.
+
+>>> op = qml.PauliX(0)
+>>> with qml.queuing.AnnotatedQueue() as q:
+...     qml.apply(op)
+...     qml.apply(op)
+>>> q.queue
+[PauliX(wires=[0]), PauliX(wires=[0])]
+
+In the case of operators composed of other operators, each operator in the composite will be queued.
+For example, both ``PauliX`` and ``PauliX`` raised to a power.
+
+>>> with qml.queuing.AnnotatedQueue() as q:
+...     base = qml.PauliX(0)
+...     pow_op = base ** 1.5
+>>> q.queue
+[PauliX(wires=[0]), PauliX(wires=[0])**1.5]
+
+In this case, each object will have metadata associated with it. At later processing steps, the original
+``PauliX`` will be ignored in favor of the operator that *owns* it.
```

Comment:
What are those later processing steps? I am slightly confused here because as we see both base and pow_op are queued but then later ignored - how? 🤔

Reply:
today, it depends on the "owner" trait as handled [here](https://github.com/PennyLaneAI/pennylane/blob/0fe04b8e20845b96bf8cc998f9be7704835994db/pennylane/queuing.py#L367). This processing function is called [whenever we exit a tape](https://github.com/PennyLaneAI/pennylane/blob/0fe04b8e20845b96bf8cc998f9be7704835994db/pennylane/tape/tape.py#L377), ie. right after constructing a tape

Reply:
Added a section on `process_queue` now that it's in the queuing module.